modules/apps/29-fee/ibc_module.go

```diff
@@ -0,0 +1,251 @@
+	// lower down in the stack. Thus, if it is not a fee version we pass the entire version string onto the underlying
+	// application.
+	// If an invalid fee version was passed, we expect the underlying application to fail on its version negotiation.
+	if cpMwVersion == types.Version {
+		im.keeper.SetFeeEnabled(ctx, portID, channelID)
+	} else {
+		// middleware versions are not the expected version for this midddleware. Pass the full version strings along,
+		// if it not valid version for any other lower middleware, an error will be returned by base application.
+		cpAppVersion = counterpartyVersion
+	}
+
+	// call underlying app's OnChanOpenTry callback with the app versions
+	appVersion, err := im.app.OnChanOpenTry(ctx, order, connectionHops, portID, channelID, chanCap, counterparty, cpAppVersion)
+	if err != nil {
+		return "", err
+	}
+
+	if im.keeper.IsFeeEnabled(ctx, portID, channelID) {
+		return channeltypes.MergeChannelVersions(types.Version, appVersion), nil
+	}
+
+	return appVersion, nil
+}
+
+// OnChanOpenAck implements the IBCModule interface
+func (im IBCModule) OnChanOpenAck(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+	counterpartyVersion string,
+) error {
+	// If handshake was initialized with fee enabled it must complete with fee enabled.
+	// If handshake was initialized with fee disabled it must complete with fee disabled.
+	cpAppVersion := counterpartyVersion
+	if im.keeper.IsFeeEnabled(ctx, portID, channelID) {
+		var cpFeeVersion string
+		cpFeeVersion, cpAppVersion = channeltypes.SplitChannelVersion(counterpartyVersion)
+
+		if cpFeeVersion != types.Version {
+			return sdkerrors.Wrapf(types.ErrInvalidVersion, "expected counterparty version: %s, got: %s", types.Version, cpFeeVersion)
+		}
+	}
+	// call underlying app's OnChanOpenAck callback with the counterparty app version.
+	return im.app.OnChanOpenAck(ctx, portID, channelID, cpAppVersion)
+}
+
+// OnChanOpenConfirm implements the IBCModule interface
+func (im IBCModule) OnChanOpenConfirm(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	// call underlying app's OnChanOpenConfirm callback.
+	return im.app.OnChanOpenConfirm(ctx, portID, channelID)
+}
+
+// OnChanCloseInit implements the IBCModule interface
+func (im IBCModule) OnChanCloseInit(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	// delete fee enabled on channel
+	// and refund any remaining fees escrowed on channel
+	im.keeper.DeleteFeeEnabled(ctx, portID, channelID)
+	err := im.keeper.RefundFeesOnChannel(ctx, portID, channelID)
+	// error should only be non-nil if there is a bug in the code
+	// that causes module account to have insufficient funds to refund
+	// all escrowed fees on the channel.
+	// Disable all channels to allow for coordinated fix to the issue
+	// and mitigate/reverse damage.
+	// NOTE: Underlying application's packets will still go through, but
+	// fee module will be disabled for all channels
+	if err != nil {
+		im.keeper.DisableAllChannels(ctx)
```

Comment:
Is it safe to just disable the fee for all channels on one chain, without notifying the counterparty chains?

According to the logic of `OnRecvPacket` and `OnAcknowledgementPacket`, whether the fee module is enabled or not depends on the local state of the chain. So what would happen if one chain has fee enabled and the other chain has fee disabled, either via `DisableAllChannels` or some other corner cases? Would it cause the raw `Acknowledgement` to be serialized and deserialized in different ways, thereby causing the deserialization to fail?